### PR TITLE
Adds log removal support

### PR DESCRIPTION
**Pull request by tinybike**
Added log removal support for Approval, Transfer, MarketCreated, OrderCanceled, OrderCreated, and OrderFilled events.  @epheph let me know you think if this looks like a reasonable approach to log removal!

### CODE REVIEW

File: src/blockchain/log-processors/order-created.ts [hunk 1]
```diff
@@ -56,8 +56,20 @@ export function processOrderCreatedLog(db: Knex, augur: Augur, trx: Knex.Transac
...
+            db.transacting(trx).from("orders").where(orderID).update(orderData).asCallback(callback);
+          }
+        });
       });
     });
   });
 }
+
+export function processOrderCreatedLogRemoval(db: Knex, augur: Augur, trx: Knex.Transaction, log: FormattedLog, callback: ErrorCallback): void {
+  processOrderCanceledLog(db, augur, trx, log, callback);
```
Comment (epheph) [inline src/blockchain/log-processors/order-created.ts]: Interesting, so we're not actually removing a created order during a reorg, we're just marking it as cancelled, but when it comes along to be recreated later, the insert-or-update logic above will do the right thing? The artifact here is that an order will exist with `isRemoved: 1`, but that seems harmless

**Comment by epheph** [comment]
Using blockstream, we can also get notified of block removals as well, what do you think of implementing that? The current `onNewBlock()` code is inelegant (using `db.raw`). I took a pass at doing a quick fix, but will need to implement real revert detection, not just `ON CONFLICT ON CONSTRAINT` or the like. Here's the code from the cleanup, (but does not handle reorgs):

https://github.com/AugurProject/augur-node/compare/sbigelow/block-table-knex?expand=1

File: src/blockchain/log-processors/order-created.ts (hunk 1 above)
Comment (tinybike) [reply src/blockchain/log-processors/order-created.ts]: Yep, that seemed to be the easiest way to deal with the whole "but if we cancel it and the cancel gets removed, we no longer have the order info so what do we do :(" problem!

File: src/blockchain/log-processors/legacy-rep-contract/transfer.ts [hunk 2]
```diff
@@ -3,15 +3,17 @@ import * as Knex from "knex";
...
+    recipient:       log.to,
+    token:           log.address,
+    value:           log.value,
+    blockNumber:     log.blockNumber,
+  }).into("transfers").asCallback(callback);
+}
 
-  db.transacting(trx).insert(dataToInsert).into("transfers").asCallback(callback);
+export function processTransferLogRemoval(db: Knex, augur: Augur, trx: Knex.Transaction, log: FormattedLog, callback: ErrorCallback): void {
+  db.transacting(trx).from("transfers").where({ transactionHash: log.transactionHash, logIndex: log.logIndex, blockNumber: log.blockNumber, token: log.address }).del().asCallback(callback);
```
Comment (epheph) [inline src/blockchain/log-processors/legacy-rep-contract/transfer.ts]: Could just use transactionHash/logIndex ?

**Comment by tinybike** [comment]
Ah, I'm not sure if BlockStream emits a special notification for blocks when reorgs happen (@MicahZoltu would prob know).  I think what we'd need to do is, if we ever see a duplicate block number (meaning that a reorg has happened), drop all blocks that are _newer_ than the duplicate block.  The other tables are all updated off of the log notifications, so they shouldn't need manual updating in this case.

File: src/blockchain/log-processors/legacy-rep-contract/transfer.ts (hunk 2 above)
Comment (tinybike) [reply src/blockchain/log-processors/legacy-rep-contract/transfer.ts]: That seems reasonable.  Is that guaranteed to be unique?

Comment (epheph) [reply src/blockchain/log-processors/legacy-rep-contract/transfer.ts]:
```CREATE TABLE transfers (
      "transactionHash" varchar(66) NOT NULL,
      "logIndex" integer NOT NULL CONSTRAINT "nonnegativeLogIndex" CHECK ("logIndex" >= 0),
      sender varchar(66) NOT NULL,
      recipient varchar(66) NOT NULL,
      token varchar(66) NOT NULL,
      value numeric NOT NULL CONSTRAINT "positiveValue" CHECK (value > 0),
      "blockNumber" integer NOT NULL CONSTRAINT "positiveBlockNumber" CHECK ("blockNumber" > 0),

      UNIQUE("transactionHash", "logIndex")

    );
```

**Comment by epheph** [comment]
@MicahZoltu and i were discussing this yesterday, it does emit block removal notifications.

**Comment by tinybike** [comment]
Oh nice!  Then it should be simple to just write an `onBlockRemoved` function that removes blocks.  Is it the same as the log notifications, where there's a `removed: true` field attached to the block, or how's that work?

**Comment by tinybike** [comment]
Yes, our types in general could use a bit of cleanup, I think there's a fair amount of duplication etc. as well.

**Pull request merged**